File: java/spring-boot-mongodb/src/test/java/com/meridal/examples/springbootmongo/mvc/RecordingControllerTest.java
package com.meridal.examples.springbootmongo.mvc;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.google.common.collect.Lists;
import com.meridal.examples.springbootmongo.domain.Recording;
import com.meridal.examples.springbootmongo.domain.Song;
import com.meridal.examples.springbootmongo.service.RecordingService;
import com.meridal.examples.springbootmongo.test.TestFramework;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;

import static org.hamcrest.Matchers.containsString;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;


@WebMvcTest
public class RecordingControllerTest extends TestFramework {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final String ID = "1234";
    private static final String OTHER_ID = "5678";

    @Autowired
    private MockMvc mvc;

    @MockBean
    private RecordingService service;

    private ObjectWriter jsonWriter;

    private Recording recording;
    private Recording otherRecording;
    private Song song;
    private Song otherSong;

    @BeforeEach
    public void setup() {
        MAPPER.configure(SerializationFeature.WRAP_ROOT_VALUE, false);
        this.jsonWriter = MAPPER.writer().withDefaultPrettyPrinter();
        this.recording = this.randomRecording();
        this.otherRecording = this.randomRecording();
        this.song = this.recording.getSongs().get(0);
        this.otherSong = this.otherRecording.getSongs().get(0);
    }

    @Test
    public void testGet() throws Exception {
        this.recording.setId(ID);
        when (this.service.getRecording(ID)).thenReturn(this.recording);
        this.mvc.perform(get("/api/recording/" + ID))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())));
    }

    @Test
    public void testGetWithInvalidID() throws Exception {
        when (this.service.getRecording(ID)).thenReturn(null);
        this.mvc.perform(get("/api/recording/" + ID))
            .andExpect(status().isNotFound());
    }

    @Test
    public void testGetAll() throws Exception {
        final List<Recording> recordings = Lists.newArrayList(this.recording, this.otherRecording);
        when (this.service.getAllRecordings()).thenReturn(recordings);
        this.mvc.perform(get("/api/recording/"))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())))
            .andExpect(content().string(containsString(this.otherRecording.getArtist())))
            .andExpect(content().string(containsString(this.otherRecording.getTitle())))
            .andExpect(content().string(containsString(this.otherRecording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.otherSong.getTitle())))
            .andExpect(content().string(containsString(this.otherSong.getDuration())));
    }

    @Test
    public void testGetAllWithIds() throws Exception {
        final List<Recording> recordings = Lists.newArrayList(this.recording, this.otherRecording);
        final List<String> ids = Lists.newArrayList(ID, OTHER_ID);
        when (this.service.getRecordings(ids)).thenReturn(recordings);
        this.mvc.perform(get("/api/recording/?ids=" + ID + "," + OTHER_ID))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())))
            .andExpect(content().string(containsString(this.otherRecording.getArtist())))
            .andExpect(content().string(containsString(this.otherRecording.getTitle())))
            .andExpect(content().string(containsString(this.otherRecording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.otherSong.getTitle())))
            .andExpect(content().string(containsString(this.otherSong.getDuration())));
    }

    @Test
    public void testPost() throws Exception {
        final String json = jsonWriter.writeValueAsString(this.recording);
        this.recording.setId(ID);
        when (this.service.saveRecording(any())).thenReturn(this.recording);
        mvc.perform(post("/api/recording/")
            .contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(status().isOk())
            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())));
    }

    @Test
    public void testPut() throws Exception {
        this.recording.setId(ID);
        final String json = jsonWriter.writeValueAsString(this.recording);
        when (this.service.saveRecording(any())).thenReturn(this.recording);
        mvc.perform(put("/api/recording/" + ID)
            .contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(status().isOk())
            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())));
    }

    @Test
    public void testDelete() throws Exception {
        when (this.service.deleteRecording(ID)).thenReturn(this.recording);
        this.mvc.perform(delete("/api/recording/" + ID))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(this.recording.getArtist())))
            .andExpect(content().string(containsString(this.recording.getTitle())))
            .andExpect(content().string(containsString(this.recording.getCatalogueNumber())))
            .andExpect(content().string(containsString(this.song.getTitle())))
            .andExpect(content().string(containsString(this.song.getDuration())));
    }

    @Test
    public void testDeleteWithInvalidID() throws Exception {
        when (this.service.deleteRecording(ID)).thenReturn(null);
        this.mvc.perform(delete("/api/recording/" + ID))
            .andExpect(status().isNotFound());
    }
}

File: kotlin/kotlin-spring-boot-mongodb/src/test/kotlin/com/meridal/examples/kotlinspringboot/controller/RecordingControllerTest.kt
package com.meridal.examples.kotlinspringboot.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.ObjectWriter
import com.fasterxml.jackson.databind.SerializationFeature
import com.meridal.examples.kotlinspringboot.model.Recording
import com.meridal.examples.kotlinspringboot.service.RecordingService
import com.meridal.examples.kotlinspringboot.test.TestFramework
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@WebMvcTest
class RecordingControllerTest(@Autowired val mockMvc: MockMvc): TestFramework() {

    private val ID = "1234"
    private val OTHER_ID = "5678"

    @MockkBean
    private lateinit var service: RecordingService
    private val mapper = ObjectMapper()
    private lateinit var jsonWriter: ObjectWriter

    private val recording = randomRecording(ID)
    private val other = randomRecording(OTHER_ID)
    private val recordings = listOf(recording, other)
    private val ids = listOf(ID, OTHER_ID)

    @BeforeEach
    fun setup() {
        mapper.configure(SerializationFeature.WRAP_ROOT_VALUE, false);
        jsonWriter = mapper.writer().withDefaultPrettyPrinter()
    }

    @Test
    fun `Get all recordings returns a list of everything in the repository`() {
        every { service.getAllRecordings() } returns recordings
        mockMvc.perform(MockMvcRequestBuilders.get("/api/recording/").accept(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.[0].artist").value(recording.artist))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.[1].artist").value(other.artist))
    }

    @Test
    fun `Get a recording by ID returns a recording`() {
        every { service.getRecording(ID) } returns recording
        mockMvc.perform(MockMvcRequestBuilders.get("/api/recording/$ID").accept(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.artist").value(recording.artist))
    }

    @Test
    fun `Get a recording by an invalid ID returns status not found`() {
        val nonexistent: Recording? = null
        every { service.getRecording(ID) } returns nonexistent
        mockMvc.perform(MockMvcRequestBuilders.get("/api/recording/$ID").accept(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isNotFound)
    }

    @Test
    fun `Post a recording places a new recording in the repository`() {
        val json = jsonWriter.writeValueAsString(recording)
        val slot = slot<Recording>()
        every { service.saveRecording(recording = capture(slot)) } returns recording
        mockMvc.perform(
            MockMvcRequestBuilders.post("/api/recording/").contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.artist").value(recording.artist))
        verify(exactly = 1) { service.saveRecording(any()) }
        Assertions.assertThat(slot.captured.artist).isEqualTo(recording.artist)
    }

    @Test
    fun `Put a recording updates a recording in the repository`() {
        val json = jsonWriter.writeValueAsString(recording)
        val slot = slot<Recording>()
        every { service.saveRecording(recording = capture(slot)) } returns recording
        mockMvc.perform(
            MockMvcRequestBuilders.put("/api/recording/$ID").contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.artist").value(recording.artist))
        verify(exactly = 1) { service.saveRecording(any()) }
        Assertions.assertThat(slot.captured.artist).isEqualTo(recording.artist)
    }

    @Test
    fun `Delete a recording by valid ID removes a recording from the repository and returns it`() {
        every { service.deleteRecording(ID) } returns recording
        mockMvc.perform(MockMvcRequestBuilders.delete("/api/recording/$ID").accept(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.jsonPath("\$.artist").value(recording.artist))
    }

    @Test
    fun `Delete a recording by an invalid ID returns status not found`() {
        val nonexistent: Recording? = null
        every { service.deleteRecording(ID) } returns nonexistent
        mockMvc.perform(MockMvcRequestBuilders.delete("/api/recording/$ID").accept(MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isNotFound)
    }
}